kpatch-build/kpatch-elf.c

```diff
@@ -206,7 +208,60 @@ void kpatch_create_rela_list(struct kpatch_elf *kelf, struct section *sec)
 			ERROR("could not find rela entry symbol\n");
 		if (rela->sym->sec &&
 		    (rela->sym->sec->sh.sh_flags & SHF_STRINGS)) {
-			rela->string = rela->sym->sec->data->d_buf + rela->addend;
+
+			/*
+			 * On s390,
+			 * Usecase 1:
+			 *
+			 * Relocation section '.rela__bug_table' at offset 0x47e0 contains 2 entries:
+			 *     0000000000000000  0000007400000005 R_390_PC32 0000000000000000 __netif_receive_skb_core.constprop.0 + d54
+			 *     0000000000000004  0000007500000005 R_390_PC32 0000000000000478 .L1^B317 + 4
+			 *
+			 * Disassembly of section .rodata.str:
+			 * 0000000000000000 <.L1^B45>:
+			 * <data>
+			 *
+			 * 0000000000000478 <.L1^B317>:
+			 *  478:   6e 65 74 2f
+			 *  47c:   63 6f 72 65
+			 *  480:   2f 64
+			 *  482:   65 76 2e 63
+			 *
+			 * echo "net/core/dev.c" | hexdump -C
+			 * 6e 65 74 2f 63 6f 72 65  2f 64 65 76 2e 63 0a
+			 *
+			 * Disassembly of section __bug_table:
+			 * 0: R_390_PC32   __netif_receive_skb_core.constprop.0+0xd54
+			 * 4: R_390_PC32   .L1^B317+0x4
+			 *
+			 * As per ABI, R_390_PC32 implies (S+A-P). The addend of
+			 * 4 just cancels out: (S+4-(4+addr of file in bug_entry))
+			 *
+			 * rela symbol L1^B317 points to the beginning of
+			 * .rodata.str section + value of the associated symbol
+			 * L1^B317. Restrict this usecase to rela__bug_table, as
+			 * LOCAL_LABEL_CHAR might appear in some other section
+			 * and might need special case handling.
+			 *
+			 * Usecase 2:
+			 *
+			 * Relocation section '.rela.text.cmdline_proc_show' at offset 0xf60:
+			 *     0000000000000018  000000430000001a R_390_GOTENT   0000000000000000 .LC1 + 2
+			 *     0000000000000030  0000006300000014 R_390_PLT32DBL 0000000000000000 seq_printf + 2
+			 *
+			 * Disassembly of section .rodata.cmdline_proc_show.str1.2:
+			 * <.LC0>:
+			 * 	...
+			 * <.LC1>:
+			 *  	...
+			 *
+			 */
+			if ((rela->type == R_390_PC32 && strstr(sec->name, ".rela__bug_table")) ||
+				(rela->type == R_390_GOTENT && strstr(rela->sym->name, ".LC")))
+				rela->string = rela->sym->sec->data->d_buf + rela->sym->sym.st_value;
+			else
+				rela->string = rela->sym->sec->data->d_buf + rela->addend;
```

Comment:
The addend of 4 really didn't make sense to me, but after staring at it for a while, I now get it.  It's a quirk of the bug table and CONFIG_GENERIC_BUG_RELATIVE_POINTERS.
```
Relocation section [214] '.rela__bug_table' for section [213] '__bug_table' at offset 0x79c78 contains 20 entries:
  Offset              Type            Value               Addend Name
  000000000000000000  390_PC32        000000000000000000    +350 .text.alloc_thread_stack_node
  0x0000000000000004  390_PC32        000000000000000000      +4 .L1^B14
```
That comes from the following code
```
#define __EMIT_BUG(x) do {					\
	asm_inline volatile(					\
		"0:	mc	0,0\n"				\
		".section .rodata.str,\"aMS\",@progbits,1\n"	\
		"1:	.asciz	\""__FILE__"\"\n"		\
		".previous\n"					\
		".section __bug_table,\"awM\",@progbits,%2\n"	\
		"2:	.long	0b-2b,1b-2b\n"			\
		"	.short	%0,%1\n"			\
		"	.org	2b+%2\n"			\
		".previous\n"					\
		: : "i" (__LINE__),				\
		    "i" (x),					\
		    "i" (sizeof(struct bug_entry)));		\
} while (0)
```
Specifically this line:
```
		"2:	.long	0b-2b,1b-2b\n"			\
```
`1b-2b` makes a PC- relative pointer to the `__FILE__` string.  Usually a PC-relative pointer is calculated starting from the address of the pointer itself.  i.e., if this were a "normal" PC32 relocation, it would be:
```
		"2:	.long	0b-2b\n"			\
		"3:	.long	1b-3b\n"			\
```
But instead, the calculation starts from the beginning of the `bug_entry` struct, which is **four** bytes before the file string pointer.  Thus the 4-byte addend.

I guess the kernel does it that way so that bug_get_file_line() works as expected for all arches.
```
  CONFIG_GENERIC_BUG_RELATIVE_POINTERS - use 32-bit pointers relative to
	the containing struct bug_entry for bug_addr and file.
```
It works, but it's still it's a weird way to do it.  Maybe I'll post a patch upstream to fix it so we can drop this change.

Reply:
I tried out the bug table fix patch. Thanks, However, I still believe, the check for NOTYPE vs SECTION would be required. Correct me if I am wrong.
 

Example usecase on s390x:
<details>
<summary>
When type is NOTYPE: -> Need to consider the rela->sym->sym.st_value
</summary>
sec->name: .rela.init.text rela->sym->name : .LC24 rela->sym->sec->name: .rodata.str1.2 rela->addend: 2 rela->sym->sym.st_value:32 rela->sym->type: 0 rela->type: 26 rela->string: sighand_cache.

</details>

<details>
<summary>
When type is STT_SECTION -> Need to consider the rela->addend.
</summary>
sec->name: .rela.data.rel.event_exit__clone rela->sym->name : .rodata.str1.2 rela->sym->sec->name: .rodata.str1.2 rela->addend: 270 rela->sym->sym.st_value:0 rela->sym->type: 3 rela->type: 22 rela->string: sys_exit_clone
</details>

Reply:
With the bug_entry fix is it possible now to combine them into a single case?
```
rela->string = rela->sym->sec->data->d_buf + rela->sym->st.value + rela->addend + rela_add_off(rela);
```

Reply:
* I think, the first approach keeps it simple. 

* I tried out the second proposed approach (Please see the attached patch):  [rela_add_off_patch.txt](https://github.com/dynup/kpatch/files/8667543/rela_add_off_patch.txt)
  - Involves code duplication in kpatch_replace_sections_syms() and rela_add_off()
  - Duplicated it because of continue statement in kpatch_replace_sections_syms().
  - Need to export rela_insn(), because it is being used in rela_add_off().
  - Also 3 args needed for rela_add_off(), As it calls rela_insn() for x86 case.

Reply:
There's no need for the code duplication.  The 'continue' is probably a bug, it should instead be an ERROR for an unhandled type.  There are several cleanups needed, I'll open a pull request.